python/lambdas/copy-files-from-tdr/lambda_function.py

```diff
@@ -0,0 +1,84 @@
+import os
+import json
+import boto3
+
+s3_client = boto3.client('s3')
+sqs_client = boto3.client('sqs')
+
+
+def lambda_handler(event, context):
+    destination_bucket = os.environ["DESTINATION_BUCKET"]
+    destination_queue = os.environ["DESTINATION_QUEUE"]
+    for record in event['Records']:
+        body = json.loads(record['body'])
+        file_id = body['fileId']
+        source_bucket = body['bucket']
+        file_location = copy_object(destination_bucket, file_id, source_bucket)
+        copy_object(destination_bucket, f"{file_id}.metadata", source_bucket)
+
+        sqs_client.send_message(QueueUrl=destination_queue, MessageBody=json.dumps({'location': file_location}))
+
+
+def copy_object(destination_bucket, s3_key, source_bucket):
+    try:
+        response = s3_client.head_object(Bucket=source_bucket, Key=s3_key)
+        file_size = response['ContentLength']
+
+        if file_size > 5 * 1024 * 1024 * 1024:
+            return multipart_copy(source_bucket, destination_bucket, s3_key, file_size)
+        else:
+            return standard_copy(source_bucket, destination_bucket, s3_key)
+
+    except Exception as e:
+        print(f"Error processing file {s3_key} from bucket {source_bucket}: {str(e)}")
+        raise e
+
+
+def standard_copy(source_bucket, destination_bucket, s3_key):
+    try:
+        copy_source = {'Bucket': source_bucket, 'Key': s3_key}
+        s3_client.copy_object(CopySource=copy_source, Bucket=destination_bucket, Key=s3_key)
+        print(f"Standard copy of {s3_key} completed successfully.")
+        return f"s3://{destination_bucket}/{s3_key}"
+    except Exception as e:
+        print(f"Error during standard copy of {s3_key}: {str(e)}")
+        raise e
+
+
+def multipart_copy(source_bucket, destination_bucket, s3_key, file_size):
+    try:
+        copy_source = {'Bucket': source_bucket, 'Key': s3_key}
+        multipart_upload = s3_client.create_multipart_upload(Bucket=destination_bucket, Key=s3_key)
+        upload_id = multipart_upload['UploadId']
+
+        part_size = 5 * 1024 * 1024
```

Comment:
It's to do with parallelism, more parts will technically process faster but you're limited by bandwidth and processing power so I doub't there's much difference in reality.

The reason we're doing it with files over 5G is because S3 doesn't support a standard copy over 5G. The advantage of using standard copy is that there's less overhead in the lambda itself and it's cheaper because you pay for every request you send to S3. So it's worth only using the multipart upload when you need to.

All that being said, 5Mb is a bit low so I think if we just aim for 100 parts by doing `file_size / 100` It's not something we can really be sure of without testing it with a load of large files which we'll do at some point and we can tweak it then.